File: more_cases/ODcaller.py
import hashlib
import os
import time
import linecache
from zokrates_pycrypto.eddsa import PrivateKey, PublicKey
from zokrates_pycrypto.field import FQ

def generate_inputs(data_len):
    # Seeded for debug purpose
    key = FQ(1997011358982923168928344992199991480689546837621580239342656433234255379025)
    sk = PrivateKey(key)

    pk = PublicKey.from_private(sk)

    inputs = [i for i in range(data_len)]

    keys = [pk.p.x.n, pk.p.y.n, sk.fe.n]
    # Fix the hash of the last block to 0x00000000000000010000000200000003000000040000000500000006...(512 bits) for debug purpose
    args = " ".join(map(str, keys)) + " " + " ".join(map(str, inputs)) + " 0 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15"
    return args


def micro_test(OD_NAME, data_len, path):
    # replace DATA_LEN
    window_size = 4
    profile_len = data_len - window_size + 1
    with open(OD_NAME + "_template.zok", "r") as f:
        data = f.read().replace("DATA_LEN", str(data_len)).replace("WINDOW_SIZE", str(window_size)).replace("PROFILE_LEN", str(profile_len))
        with open(path + OD_NAME + ".zok", "w+") as file:
            file.write(data)

    print("data length: %s" % data_len)

    # compile circuit
    os.system("cd " + path + " && zokrates compile -i " + OD_NAME + ".zok -o " + OD_NAME + " > compile.out")
    linecache.updatecache(path + "compile.out")
    constraints = linecache.getline(path + "compile.out", 4).split(":")[1].strip()
    print("# of constraints: %s" % constraints)

    # GenParam
    st = time.perf_counter()
    os.system("cd " + path + " && zokrates setup -i " + OD_NAME + " > setup.out")
    ed = time.perf_counter()
    linecache.updatecache(path + "setup.out")
    points = linecache.getline(path + "setup.out", 3).split(" ")[2].strip()
    print("# of points: %s" % points)
    print("GenParam: %s ms" % ((ed - st) * 1000))
    genparam = (ed - st) * 1000

    # ProveData
    args = generate_inputs(data_len)
    st = time.perf_counter()
    os.system("cd " + path + " && zokrates compute-witness -i " + OD_NAME + " -a " + args + " > witness.out && zokrates generate-proof -i " + OD_NAME + " > proof.out")
    ed = time.perf_counter()
    print("ProveData: %s ms" % ((ed - st) * 1000))
    provedata = (ed - st) * 1000

    # VerifyProof
    st = time.perf_counter()
    os.system("cd " + path + " && zokrates verify > verify.out")
    # 512bits -> 64 bytes, simulate the operation of calculating hash of the last block, '8e6245e107a0127f17e480ba65f27e20ac48d13f15eedc93b716eb2806701f7d'
    sha2 = hashlib.sha256(b'0000000100020003000400050006000700080009001000110012001300140015').hexdigest()
    ed = time.perf_counter()
    print("VerifyProof: %s ms\n" % ((ed - st) * 1000))
    verifyproof = (ed - st) * 1000

    logs = [data_len, constraints, points, genparam, provedata, verifyproof]
    return ",".join(map(str, logs))


if __name__ == "__main__":
    OD_NAME = "MP"
    f = open(OD_NAME + "_micro_test.csv", "w+")
    f.write("data_len,constraints,points,genparam,provedata,verifyproof\n")
    if not os.path.exists("build"):
        os.makedirs("build")
    for data_len in range(50, 601, 50):
        path = "./build/" + str(data_len) + "/"
        if not os.path.exists(path):
            os.makedirs(path)
        f.write(micro_test(OD_NAME, data_len, path) + "\n")
    f.close()
    

File: more_cases/ODcaller.js
const { initialize } = require('zokrates-js')
const fs = require("fs");

// ecc key pair for test purpose [pk.p.x.n, pk.p.y.n, sk.fe.n]
const keys = ["14897476871502190904409029696666322856887678969656209656241038339251270171395", "16668832459046858928951622951481252834155254151733002984053501254009901876174", "1997011358982923168928344992199991480689546837621580239342656433234255379025"];

// dummy block hash (512 bits)
const blockhash = ["0", "1", "2", "3", "4", "5", "6", "7", "8", "9", "10", "11", "12", "13", "14", "15"];

const DEBUG = false;

initialize().then((zokratesProvider) => {
    for (let DATA_LEN = 50; DATA_LEN <= 50; DATA_LEN += 50) {
        console.info("====================\nDATA_LEN: " + DATA_LEN.toString() + "\n====================");

        // dummy inputs
        let input = new Array(DATA_LEN);
        for (var i = 0; i < DATA_LEN; i++) {
            input[i] = i.toString();
        }

        let zok_file = "MP.zok";
        const source = fs.readFileSync(zok_file).toString().replaceAll("DATA_LEN", DATA_LEN.toString()).replaceAll("WINDOW_SIZE", "4").replaceAll("PROFILE_LEN", (DATA_LEN-4+1).toString());
        if (DEBUG) {
            console.log("source code:\n" + source);
        }

        // compile circuit
        console.time("compile circuit");
        const artifacts = zokratesProvider.compile(source);
        console.timeEnd("compile circuit");
        if (DEBUG) {
            console.log("compiled artifacts:\n" + artifacts.program);
        }

        // GenParam
        console.time("GenParam");
        const keypair = zokratesProvider.setup(artifacts.program);
        console.timeEnd("GenParam");

        // ProveData
        let args = [keys.slice(0,2), keys[2], input, blockhash];
        if (DEBUG) {
            console.log("args:\n" + args);
        }

        console.time("ProveData");
        const { witness, output } = zokratesProvider.computeWitness(artifacts, args);
        const proof = zokratesProvider.generateProof(artifacts.program, witness, keypair.pk);
        console.timeEnd("ProveData");
        if (DEBUG) {
            console.log("output:\n" + output);
        }

        // export solidity verifier
        // const verifier = zokratesProvider.exportSolidityVerifier(keypair.vk);
        // console.log(verifier);
        
        // VerifyProof
        console.time("VerifyProof");
        const isVerified = zokratesProvider.verify(keypair.vk, proof);
        console.timeEnd("VerifyProof");
        if (DEBUG) {
            console.log("isVerified:\n" + isVerified.toString());
        }
    }
});
